src/archive.rs

```diff
@@ -394,26 +396,65 @@ impl<'a> EntriesFields<'a> {
                 if let Some(pax_extensions_ref) = &pax_extensions {
                     pax_size = pax_extensions_size(pax_extensions_ref);
                 }
+                // Not an entry
+                // Keep pax_extensions for the next ustar header
+                processed -= 1;
                 continue;
             }
 
             let mut fields = EntryFields::from(entry);
+            fields.pax_extensions = pax_extensions;
+            pax_extensions = None; // Reset pax_extensions after use
+            if is_recognized_header && fields.is_pax_sparse() {
+                gnu_longname = fields.pax_sparse_name();
+            }
             fields.long_pathname = gnu_longname;
             fields.long_linkname = gnu_longlink;
-            fields.pax_extensions = pax_extensions;
             self.parse_sparse_header(&mut fields)?;
             return Ok(Some(fields.into_entry()));
         }
     }
 
     fn parse_sparse_header(&mut self, entry: &mut EntryFields<'a>) -> io::Result<()> {
-        if !entry.header.entry_type().is_gnu_sparse() {
+        if !entry.is_pax_sparse() && !entry.header.entry_type().is_gnu_sparse() {
             return Ok(());
         }
-        let gnu = match entry.header.as_gnu() {
-            Some(gnu) => gnu,
-            None => return Err(other("sparse entry type listed but not GNU header")),
-        };
+        let mut sparse_map = Vec::<SparseEntry>::new();
```

Comment:
One of the main goals I tried to keep for the `tar` crate is to minimize internal allocations. Instead of having a temporary list here could this be refactored to avoid the intermediate allocation?

Reply:
I don't think so. Here we need to convert strings to numbers and the number of pairs is not fixed.